tcp-native/src/main/java/org/ballerinalang/stdlib/tcp/Dispatcher.java

```diff
@@ -0,0 +1,123 @@
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+
+/**
+ * Dispatch async methods.
+ */
+public class Dispatcher {
+
+    private static final Logger log = LoggerFactory.getLogger(Dispatcher.class);
+
+    private static void invokeOnBytes(TcpService tcpService, ByteBuf buffer, Channel channel) {
+        try {
+            Object[] params = getOnBytesSignature(buffer, channel);
+
+            tcpService.getRuntime().invokeMethodAsync(tcpService.getConnectionService(), Constants.ON_BYTES, null, null,
+                    new TcpCallback(), params);
+        } catch (BError e) {
+            Dispatcher.invokeOnError(tcpService, e.getMessage());
+        }
+    }
+
+    public static void invokeOnError(TcpService tcpService, String message) {
+        try {
+            Object params[] = getOnErrorSignature(message);
+            tcpService.getRuntime().invokeMethodAsync(tcpService.getConnectionService(), Constants.ON_ERROR, null, null,
+                    new TcpCallback(tcpService), params);
+        } catch (Throwable t) {
+            log.error("Error while executing onError function", t);
+        }
+    }
+
+    private static Object[] getOnBytesSignature(ByteBuf buffer, Channel channel) {
+        byte[] byteContent = new byte[buffer.readableBytes()];
+        buffer.readBytes(byteContent);
+        BArray bytes = ValueCreator.createArrayValue(byteContent);
+        return new Object[]{bytes, true};
+    }
+
+    private static Object[] getOnErrorSignature(String message) {
+        return new Object[]{Utils.createSocketError(message), true};
+    }
+
+    private static BObject createClient(Channel channel, TcpService tcpService) {
+        InetSocketAddress remoteAddress = (InetSocketAddress) channel.remoteAddress();
+        InetSocketAddress localAddress = (InetSocketAddress) channel.localAddress();
+        final BObject caller = ValueCreator.createObjectValue(Utils.getTcpPackage(), Constants.CALLER);
+        caller.set(StringUtils.fromString(Constants.CALLER_REMOTE_PORT), remoteAddress.getPort());
+        caller.set(StringUtils.fromString(Constants.CALLER_REMOTE_HOST),
+                StringUtils.fromString(remoteAddress.getHostName()));
+        caller.set(StringUtils.fromString(Constants.CALLER_LOCAL_PORT), localAddress.getPort());
+        caller.set(StringUtils.fromString(Constants.CALLER_LOCAL_HOST),
+                StringUtils.fromString(localAddress.getHostName()));
+        caller.addNativeData(Constants.CHANNEL, channel);
+        caller.addNativeData(Constants.SERVICE, tcpService);
+        return caller;
+    }
+
+    public static void invokeRead(TcpService tcpService, ByteBuf buffer, Channel channel) {
+        for (MethodType method : tcpService.getConnectionService().getType().getMethods()) {
+            switch (method.getName()) {
+                case Constants.ON_BYTES:
+                    Dispatcher.invokeOnBytes(tcpService, buffer, channel);
+                    break;
+            }
```

Comment:
In the default shall we print error saying this can be happening.

Reply:
```  switch (method.getName()) {
                case Constants.ON_BYTES:
                    Dispatcher.invokeOnBytes(tcpService, buffer, channel);
                    break;
                case Constants.ON_ERROR: break;
                case Constants.ON_CLOSE: break;
                default: log.error("this method can't be dispatched.");
            }```

Reply:
keeping as it is for now